### PR TITLE
Add explicit null checks on ObjectPool.Return rather than triggering random failures later on

### PR DESCRIPTION
**Pull request by geeknoid**
Without these checks, it's possible to call ObjectPool.Return and pass null, which manifests itself as a later call to ObjectPool.Get returning null. This is surprising to customers and potentially hard to understand/diagnose. The null checks fix this.

### CODE REVIEW

File: src/ObjectPool/ObjectPool.sln [hunk 1]
```diff
@@ -0,0 +1,42 @@
+﻿
```
Comment (BrennanConroy) [inline src/ObjectPool/ObjectPool.sln:1]: Remove/replace. We use slnf in this repository. e.g. https://github.com/dotnet/aspnetcore/blob/main/src/Http/HttpAbstractions.slnf

Comment (geeknoid) [reply src/ObjectPool/ObjectPool.sln:1]: Whoops, that slipped in...

File: src/ObjectPool/src/DefaultObjectPool.cs [hunk 2]
```diff
@@ -66,6 +67,7 @@ public override T Get()
     /// <inheritdoc />
     public override void Return(T obj)
     {
+        ArgumentNullThrowHelper.ThrowIfNull(obj);
```
Comment (gfoidl) [inline src/ObjectPool/src/DefaultObjectPool.cs:70]: Can `null` just be a nop instead of throwing?

For diagnostics fail early may seem better, perf-wise maybe just ignore `null` may be faster?!

**Review by gfoidl — COMMENTED** [review]
Tests should be added to https://github.com/dotnet/aspnetcore/blob/12e4aac6306a13afd3d206cb41d1bbccc1b50e7d/src/ObjectPool/test/DefaultObjectPoolTest.cs#L9 too.

File: src/ObjectPool/src/DefaultObjectPool.cs (hunk 2 above)
Comment (geeknoid) [reply src/ObjectPool/src/DefaultObjectPool.cs:70]: If app code is passing null here and is worried about the overhead of the exception being thrown, then I think the app code can have a null check on their end before calling the API.

Comment (xiaoyuvax) [reply src/ObjectPool/src/DefaultObjectPool.cs:70]: do u note that NET7 does not behave this way?
i prefer nop anyway, otherwise i would seek to implement a custom objectpoolpolicy to sovle this. why not just keep it simple?

File: src/ObjectPool/src/ObjectPool.cs [hunk 3]
```diff
@@ -19,6 +19,7 @@ public abstract class ObjectPool<T> where T : class
     /// Return an object to the pool.
     /// </summary>
     /// <param name="obj">The object to add to the pool.</param>
+    /// <exception cref="System.ArgumentNullException">when <paramref name="obj"/> is null.</exception>
```
Comment (JamesNK) [inline src/ObjectPool/src/ObjectPool.cs:22]:
```suggestion
    /// <exception cref="System.ArgumentNullException"><paramref name="obj"/> is <see langword="null"/>.</exception> 
```
Matches docs like this [elsewhere](https://github.com/search?q=repo%3Adotnet%2Fruntime%20%3Cexception%20cref%3D%22ArgumentNullException%22%3E&type=code).

**Review by JamesNK — COMMENTED** [review]
Seems ok to me.

Throwing the error is a behavior change, but most likely it is changing where the error is thrown (error returning null vs error trying to use null from Get.

File: src/ObjectPool/src/DefaultObjectPool.cs (hunk 2 above)
Comment (amcasey) [reply src/ObjectPool/src/DefaultObjectPool.cs:70]: I generally agree with the argument that passing null to `Return` indicates a bug in the caller (you didn't get null from the pool, so why would you be returning it?) and that alerting callers to such bugs is generally helpful.  On the other hand, the 7.0 implementation does appear to have gracefully accepted null and just ignored it in subsequent `Get` calls.  I think I'd want to see evidence that the evidence could actually prevent real problems (i.e. in consuming programs) before I'd accept a breaking change.

TL;DR: It's a close call but I'd lean towards noop, rather than throw.

Comment (xiaoyuvax) [reply src/ObjectPool/src/DefaultObjectPool.cs:70]: > I generally agree with the argument that passing null to `Return` indicates a bug in the caller (you didn't get null from the pool, so why would you be returning it?) and that alerting callers to such bugs is generally helpful. 

The life cycle of a pooled object should not necessarily be a closed loop, some new instances might be fed to return() or some references to old instances might also evolve to be null with the instance disposed, as to decouple the Get/Return lifecycle management.  Coupling'em is a loose choice rather than a tight must.

otherwise, as a practical tool, objectpool should be designed to work as smoothy as possible while considering performance/gracefullness the sametime, instead of asking for unncessary care from programmers from time to time, for it is still not quite efficient/convenient to use.
For complicated cases, such as with multi-references or non-synchronizable life-cycles of referenced underlying children objects, it's a big challege to manange/organize objectpools properly, as not to return a still referenced/already disposed(i.e. null) object, since no way to determine referencec count (which is the fundamental problem for using objectpool), this is the biggest uncertainty.  
So maxmize the torlerance of uncertainty of objectpool is a good principle, by swallowing null checks silently is just one small dent, in my point of view.
